dvc/remote/gdrive/oauth2.py

```diff
@@ -0,0 +1,105 @@
+from hashlib import md5
+import datetime
+import json
+import os
+
+from google_auth_oauthlib.flow import InstalledAppFlow
+from google.auth.transport.requests import AuthorizedSession
+import google.oauth2.credentials
+
+from dvc.lock import Lock
+from dvc.config import Config
+
+
+class OAuth2(object):
+
+    DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S"
+
+    def __init__(self, credentialpath, scopes, config):
+        self.credentialpath = credentialpath
+        self.scopes = scopes
+        self.config = config
+
+    def get_session(self):
+        creds_storage, creds_storage_lock = self._get_storage_lock()
+        with creds_storage_lock:
+            if os.path.exists(creds_storage):
+                creds = self._load_credentials(creds_storage)
+            else:
+                creds = self._acquire_credentials()
+                self._save_credentials(creds_storage, creds)
+        return AuthorizedSession(creds)
+
+    def _get_creds_id(self, client_id):
+        plain_text = "|".join([client_id] + self.scopes)
+        hashed = md5(plain_text.encode("ascii")).hexdigest()
+        return hashed
+
+    def _acquire_credentials(self):
+        # Create the flow using the client secrets file from the
+        # Google API Console.
+        flow = InstalledAppFlow.from_client_secrets_file(
+            self.credentialpath, scopes=self.scopes
+        )
+        flow_runner = self.config.get(
+            Config.SECTION_OAUTH2_FLOW_RUNNER, "console"
+        )
+        if flow_runner == "local":
+            creds = flow.run_local_server()
+        elif flow_runner == "console":
+            creds = flow.run_console()
+        else:
+            raise Exception(
```

Comment:
raise something meaningful here

Reply:
we should be also checking this on the config level and don't allow users to set it, raise a ConfigException if something strange is set

Reply:
Great point @shcheklein . @ei-grad Let's use config SCHEMA for that to check for choices. See how link type is checked there as an example.